EQL: Adds an ability to start an asynchronous EQL search

imotov opened the pull request:
Adds support for async searches to eql search API. This commit is limited to
only submitting search API requests and doesn't provide APIs to get results
nor delete the results. These functions will be added in follow up PRs.

Relates to #49638

imotov committed:
EQL: Adds an ability to start an asynchronous EQL search

Adds support for async searches to eql search API. This commit is limited to
only submitting search API requests and doesn't provide APIs to get results
nor delete the results. These functions will be added in follow up PRs.

Relates to #49638

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/EQL)

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -5,54 +5,225 @@
+            } else {
+                // We finished after timeout - saving exception
+                storeResults(searchTask, new EqlStoredResponse(e, threadPool.absoluteTimeInMillis() + keepAlive.getMillis()));
+            }
+        });
+    }
+
+    private void storeResults(EqlSearchTask searchTask, EqlStoredResponse storedResponse) {
+        storeResults(searchTask, storedResponse, ActionListener.wrap(r -> {
+        }, e -> {
+        }));
+    }
+
+    private void storeResults(EqlSearchTask searchTask, EqlStoredResponse storedResponse, ActionListener<Void> finalListener) {
+        try {
+            asyncTaskIndexService.createResponse(searchTask.getAsyncExecutionId().getDocId(),
+                threadPool.getThreadContext().getHeaders(), storedResponse, ActionListener.wrap(
+                    // We should only unregister after the result is saved
+                    resp -> {
+                        taskManager.unregister(searchTask);
+                        logger.trace(() -> new ParameterizedMessage("stored eql search results for [{}]",
+                            searchTask.getAsyncExecutionId().getEncoded()));
+                        finalListener.onResponse(null);
+                    },
+                    exc -> {
+                        taskManager.unregister(searchTask);
+                        Throwable cause = ExceptionsHelper.unwrapCause(exc);
+                        if (cause instanceof DocumentMissingException == false &&
+                            cause instanceof VersionConflictEngineException == false) {
+                            logger.error(() -> new ParameterizedMessage("failed to store eql search results for [{}]",
+                                searchTask.getAsyncExecutionId().getEncoded()), exc);
+                        }
+                        finalListener.onFailure(exc);
+                    }));
+        } catch (Exception exc) {
+            taskManager.unregister(searchTask);
+            logger.error(() -> new ParameterizedMessage("failed to store eql search results for [{}]",
+                searchTask.getAsyncExecutionId().getEncoded()), exc);
+
+        }
```
I imagine these utility methods would be useful for every plugin doing async work, as most of the code is around the taskIndexService and taskManager hence why I see value in externalizing them.
It could be in a follow-up PR.
It would be nice to just reuse the boilerplate code (similar to say `ActionListener.wrap`) and only pass in the types either as lambdas or wrapped objects:

`storeResults(asyncTaskIndexService, searchTask, finalListener, task -> searchTask.getAsyncExecutionId().getEncoded())` 

This could sit next to `AsyncExecutionId` in the `async` module so `EqlSearchTask` would just import it and the rest of utilities (like `storeResults`) would just require suppliers for it; then the method above would be even smaller:

`storeResults(asyncTaskIndexService, searchTask, finalListener, task::getAsyncExecutionId)`

costin reviewed (commented):
My main comment is on generalizing the async infrastructure bits and promoting them as utilities inside the `xpack.core.async`. I would imagine any consumer of said API would have to go through the same process of storing and retrieving the results, having consistent parameters (`wait_for_*`).
This would also apply to testing which is focused on the async nature.

imotov committed:
Address review comments

Moves most of the logic from TransportEqlSearchAction to
AsyncTaskManagementService

imotov committed:
Merge remote-tracking branch 'elastic/master' into async-eql

imotov committed:
Merge remote-tracking branch 'elastic/master' into async-eql

javanna commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskManagementService.java:
```diff
@@ -0,0 +1,236 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class AsyncTaskManagementService<
```
What is the purpose of this service? It looks to me like a simplified fork of what async search does when it comes to waiting for completion and storing the response. Yet I see only one user of these new abstractions. Are there going to be more users of this? Async search will not be migrated to it, right?

imotov replied:
We can migrate async search here in the future. The main difference is that this service relies on cascading cancellation and async search is preforming cancellation internally. So, some pieces of this logic are missing here.

javanna replied:
Cancellation is not a problem, we need to migrate async search to rely on cascading cancellation anyways. But the logic around storing partial responses and wait for completion is very different from what you have now.

imotov replied:
I think logic is pretty much the same but implementation is quite a bit different. Removing cancellation concerns allowed me to remove all the thread management and client calls from the task into the service and keep the task lightweight. I don't think a task should be a control structure or use clients. It should be just a simple light weight data transfer structure.

javanna replied:
One point where the behaviour differs at first glance is that in async search, we never want to return nor store a response before the initial onListShards listener callback. Hence we register a completion listener, but that is not invoked until onListShards is called.
What I am saying is that there are subtle differences that are independent from task cancellation, and I am not sure that async search can be migrated to this simplified version of async management service.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchResponse.java:
```diff
@@ -32,43 +33,56 @@
     private final boolean isTimeout;
+    private final String asyncExecutionId;
+    private final boolean isRunning;
+    private final boolean isPartial;
```
is is_partial needed in eql given that no partial results are returned? We are regretting to have added this flag in the response of async search, and we will likely remove it, so it would be good to make sure that it's is needed in eql.

javanna reviewed (commented):
I left two questions. My main question is whether the new abstractions that are introduced with this PR will be shared with other async API. I worry that it may be too early to go through the exercise of generalizing as the async behaviour of eql is not consolidated yet and it seems to differ from what e.g. async search does.

imotov commented:
> My main comment is on generalizing the async infrastructure bits and promoting them as utilities inside the xpack.core.async. I would imagine any consumer of said API would have to go through the same process of storing and retrieving the results, having consistent parameters (wait_for_*).
This would also apply to testing which is focused on the async nature.

> My main question is whether the new abstractions that are introduced with this PR will be shared with other async API. I worry that it may be too early to go through the exercise of generalizing as the async behaviour of eql is not consolidated yet and it seems to differ from what e.g. async search does.

I cannot win this game :smile:. @javanna , @costin :  Should I undo my last changes and merge it back into the EQL transport?

imotov committed:
Merge remote-tracking branch 'elastic/master' into async-eql

javanna commented:
I am not against abstractions and sharing code once there is the need, but I don't see what the direction is here. Abstractions add complexity and we have only one use of those at the moment, and I would like to understand what the plan is around having more users of these new abstractions. Ideally, I would delay abstracting things until there is concrete need for it. I am saying this especially looking at how in theory EQL async aspects should share behaviour with async search although in practice they do only partially.

imotov commented:
Talked about it offline. We came up with the following plan:

In this PR:
- since we are not sure about reuse, we are going to move the logic of AsyncTaskManagementService inside EQL branch
- we will get rid of a separate index and security user for eql and reuse async search one and start reusing the same security user in EQL until we can switch to system indices

In the follow up PR:
- we will move TransportGetAsyncSearchAction and TransportDeleteAsyncSearchAction to core and make it reusable

@jimczi, @javanna , @costin please add if I missed something.

imotov commented:
Closing. Will reopen it against a branch
